plugins/catalog/src/state/types.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity } from '@backstage/catalog-model';
+
+export type EntityFilterType = 'field' | 'owned' | 'starred' | 'all';
+
+export type EntityFilter = {
+  type: EntityFilterType;
+  field?: string; // e.g. spec.type
+  values?: string[];
+};
```

Comment:
i was actually wondering if the state could hold only two things related to the filtering as such,

```ts
serverFilters: Record<string, ServerFilter>; // keyed by filter control id
clientFilters: Record<string, ClientFilter>; // keyed by filter control id
```

and then

```ts
type ServerFilter = Record<string, string[]>; // e.g { 'spec.type': ['website'] }
type ClientFilter = Array<(entity: Entity) => boolean>; // just matcher functions
```

which, whenever something changes somewhere, get merged into one big server filter (triggering a reload if the stringified version differs from last time) and one big client filter.

But even with only client filters to begin with. I was thinking that maybe the central state doesn't have to have any notion at all about what filtering options exist.

This does not take into account the calculation of counts on each control. But I'm thinking that maybe we should just abandon those, it's a pretty hard problem to get right given both server- and client-side filtering. And, like, what would they even show on checkboxes - etc.

Reply:
I can definitely create the pre-emptive split between client/server filters. Was thinking along the same lines of the server filter results being the base for the client side ones. 

It would be nice to have no notion of the filter option types. I wanted to avoid passing functions into the store state as in my books that's a no no as it can't be serialized. I have considered maybe splitting the filter definition and the filter values in 2. So perhaps the reducer is then created via a factory that can inject custom registered filter definitions. Then applying the filter is just a matter of matching the filter id to the id of the definition which contains the filter predicate function.

As for the counts, I put them in, but they are broken as I've also found it, it's tricky to get them right as they need to factor in various subsets that are not consistent with the full filter set. I'd be perfectly happy removing them 😄 .